libsolidity/formal/CHC.cpp

```diff
@@ -1175,86 +1136,91 @@ void CHC::addVerificationTarget(ASTNode const* _scope, VerificationTarget::Type
-
-				satMsgUnderflow = "Underflow (resulting value less than " + formatNumberReadable(intType->minValue()) + ")";
-				satMsgOverflow = "Overflow (resulting value larger than " + formatNumberReadable(intType->maxValue()) + ")";
-				if (target.type == VerificationTarget::Type::Underflow)
+				string satMsg;
+				string satMsgUnderflow;
+				string satMsgOverflow;
+				string unknownMsg;
+				ErrorId errorReporterId;
+				ErrorId underflowErrorId = 3944_error;
+				ErrorId overflowErrorId = 4984_error;
+
+				if (target.type == VerificationTarget::Type::PopEmptyArray)
+				{
+					solAssert(dynamic_cast<FunctionCall const*>(scope), "");
+					satMsg = "Empty array \"pop\" detected here.";
+					unknownMsg = "Empty array \"pop\" might happen here.";
+					errorReporterId = 2529_error;
+				}
+				else if (
+					target.type == VerificationTarget::Type::Underflow ||
+					target.type == VerificationTarget::Type::Overflow
+				)
 				{
-					satMsg = satMsgUnderflow + " happens here.";
-					unknownMsg = satMsgUnderflow + " might happen here.";
-					errorReporterId = underflowErrorId;
+					auto const* expr = dynamic_cast<Expression const*>(scope);
+					solAssert(expr, "");
+					auto const* intType = dynamic_cast<IntegerType const*>(expr->annotation().type);
+					if (!intType)
+						intType = TypeProvider::uint256();
+
+					satMsgUnderflow = "Underflow (resulting value less than " + formatNumberReadable(intType->minValue()) + ")";
+					satMsgOverflow = "Overflow (resulting value larger than " + formatNumberReadable(intType->maxValue()) + ")";
+					if (target.type == VerificationTarget::Type::Underflow)
+					{
+						satMsg = satMsgUnderflow + " happens here.";
+						unknownMsg = satMsgUnderflow + " might happen here.";
+						errorReporterId = underflowErrorId;
+					}
+					else if (target.type == VerificationTarget::Type::Overflow)
+					{
+						satMsg = satMsgOverflow + " happens here.";
+						unknownMsg = satMsgOverflow + " might happen here.";
+						errorReporterId = overflowErrorId;
+					}
 				}
-				else if (target.type == VerificationTarget::Type::Overflow)
+				else if (target.type == VerificationTarget::Type::DivByZero)
 				{
-					satMsg = satMsgOverflow + " happens here.";
-					unknownMsg = satMsgOverflow + " might happen here.";
-					errorReporterId = overflowErrorId;
+					satMsg = "Division by zero happens here.";
+					unknownMsg = "Division by zero might happen here.";
+					errorReporterId = 4281_error;
 				}
-			}
-			else if (target.type == VerificationTarget::Type::DivByZero)
-			{
-				satMsg = "Division by zero happens here.";
-				unknownMsg = "Division by zero might happen here.";
-				errorReporterId = 4281_error;
-			}
-			else
-				solAssert(false, "");
+				else
+					solAssert(false, "");
 
-			auto it = m_errorIds.find(scope->id());
-			solAssert(it != m_errorIds.end(), "");
-			unsigned errorId = it->second;
+				auto it = m_errorIds.find(scope->id());
+				solAssert(it != m_errorIds.end(), "");
```

Comment:
These two are the ones I meant. They could be moved before `for (size_t i = 0; i < targets.size(); ++i)`

Reply:
Ah, right. But that's tricky. That condition will actually not be true for the Assertion verification targets, so I cannot move it there.

Reply:
Ah right, because it's in the `if`. Ok, let's keep it like this then